File: framework/src/main/java/com/linxiao/framework/net/ApiResponseParseFunction.java
package com.linxiao.framework.net;

import com.linxiao.framework.json.GsonParser;

import java.lang.reflect.Type;

import io.reactivex.functions.Function;
/**
 * flatMap function to parse data in ApiResponse to data object
 *
 * @author linxiao
 * @since 2018-08-20
 */
public class ApiResponseParseFunction<T> implements Function<ApiResponse, T> {

    private final Type type;

    public ApiResponseParseFunction() {
        type = Object.class;
    }

    public ApiResponseParseFunction(Class<T> clazz) {
        type = clazz;
    }

    public ApiResponseParseFunction(Type type) {
        this.type = type;
    }

    @SuppressWarnings("unchecked")
    @Override
    public T apply(ApiResponse apiResponse) throws Exception {
        T ret;
        if (!apiResponse.isSuccess()) {
            throw new ApiResponse.ApiException(apiResponse);
        }
        if (type == Object.class) {
            return (T) (apiResponse.data == null ? "" : apiResponse.data);
        }
        ret = GsonParser.getParser().fromJson(apiResponse.data, type);
        if (ret == null) {
            throw new ApiResponse.ApiException(apiResponse);
        }
        return ret;
    }
}

File: framework/src/main/java/com/linxiao/framework/net/ApiResponseParseFunction.kt
package com.linxiao.framework.net

import com.linxiao.framework.json.GsonParser.parser
import io.reactivex.functions.Function
import java.lang.reflect.Type

/**
 * flatMap function to parse data in ApiResponse to data object
 *
 * @author linxiao
 * @since 2018-08-20
 */
@Suppress("UNCHECKED_CAST")
class ApiResponseParseFunction<T> : Function<ApiResponse, T> {
    private val type: Type

    constructor() {
        type = Any::class.java
    }

    constructor(clazz: Class<T>) {
        type = clazz
    }

    constructor(type: Type) {
        this.type = type
    }

    @Throws(Exception::class)
    override fun apply(apiResponse: ApiResponse): T {
        if (type === Any::class.java) {
            return (if (apiResponse.data == null) "" else apiResponse.data) as T
        }
        return parser.fromJson(apiResponse.data, type) ?: throw ApiException(apiResponse)
    }
}